### ISSUE TITLE
pysat.card.CardEnc.atleast tries to set signal handler

### ISSUE THREAD

**Issue by rjungbeck** [pysathq/pysat#28]
I want to do all solver related work in it's  own thread (to keep the GUI responsive). 

This includes CNF creation (because it could have several million clauses).

But setting a signal handler only works in the main thread (and raises a ValueError exception in otzher threads).

Why does  pysat.card.CardEnc.equals (encoding=pysat.card.EncType.sortnetwrk) (that ultimately calls atleast) needs to set the signal handler at all? It should just create some CNFclauses.

The simplest solution would probably be to have a try/except around the signal call that ignore the error.

**Comment by alexeyignatiev** [comment]
Hi @rjungbeck,

The routines of most C extensions included as modules (`card` and `solvers`) indeed set a signal handler. The rationale is simply to be able to stop a script that uses PySAT if it takes too much time. Concretely, `pysat.card.CardEnc.equals` (and the other similar methods of module `card`) may take time if they are instructed to build a huge cardinality encoding. (As you mentioned, a CNF to construct can have millions of clauses.) As a user, I want to stop the script under some circumstances. Without the signal handler, it is problematic.

Now, I am not sure I understand what you suggest to solve the problem. Can you elaborate?

**Comment by rjungbeck** [comment]
I understand your desire. But it only works for single threaded programs.
I currently use the following (monkey patching) solution:

````
import signal
origSignal=None

def solveThread().
    
    global origSignal

    def DummySignal(*args):
        try:
            return origSignal(*args)
        except ValueError:
            return None

    if origSignal==None:
        origSignal=signal.signal
        signal.signal=DummySignal

    createCnf()
````
It effectively replaces signal.signal with a nop when it is called outside of the main thread.

**Comment by alexeyignatiev** [comment]
Hi @rjungbeck,

OK, I made an attempt to fix this issue in [1540cfe](https://github.com/pysathq/pysat/commit/1540cfe86aef59223fd4eb06bd022056c493731b). The version of the toolkit is also bumped to 0.1.4.dev24.

As I am not an expert in these things, please, check if the issue is resolved in your case.

**Comment by rjungbeck** [comment]
Yes, your solution works (at least for me).
